scripts/globals/job_utils/beastmaster.lua

```diff
@@ -0,0 +1,132 @@
+    local extraDurationFromMod = charmDuration * (charmTimeMod * 0.5 / 10) -- Assumes 5% per charmTimeMod
+    charmDuration = charmDuration + extraDurationFromMod
+
+    return math.floor(charmDuration)
+end
+
+xi.job_utils.beastmaster.getCharmChance = function(charmer, target, includeMods)
+    if
+        not charmer or                                -- Invalid charmer
+        not target or                                 -- Invalid target
+        not charmer:isPC() or                         -- Charmer not a player
+        not target:isMob() or                         -- Target not a mob
+        target:getMobMod(xi.mobMod.CHARMABLE) == 0 or -- Not charmable
+        (target:getMaster() ~= nil)                   -- Someone else's pet
+    then
+        return 0
+    end
+
+    -- Use the players BST level (even if subjob) for charm chance calc
+    local charmerJobLevel = charmer:getJobLevel(xi.job.BST)
+    local targetLevel     = target:getMainLvl()
+    local charmres        = target:getMod(xi.mod.CHARMRES)
+    local charmChance     = 50 - charmres
+    -- dLvl only applies when player lvl < mob lvl
+    -- and varies for different target levels
+    if charmerJobLevel < targetLevel
+        if targetLevel >= 71 then
+            charmChance = charmChance - 10 * (targetLevel - charmerJobLevel)
+        elseif targetLevel >= 51 then
+            charmChance = charmChance - 5 * (targetLevel - charmerJobLevel)
+        else
+            charmChance = charmChance - 3 * (targetLevel - charmerJobLevel)
+        end
+    end
+
+    -- Another multiplier determined by target light res rank
+    -- as charm is a light based ability
+    local rank = target:getMod(xi.mod.LIGHT_RES_RANK)
+    if rank <= -3 then
+        charmChance = charmChance * 1.5
+    elseif rank <= -2 then
+        charmChance = charmChance * 1.4
+    elseif rank <= -1 then
+        charmChance = charmChance * 1.2
+    elseif rank <= 0 then
+        charmChance = charmChance
+    else
+        charmChance = charmChance / 2
+    end
+
+    -- Need a includeMods param because staves (which give CHARM_CHANCE) are not taken into account for Gauge
+    if includeMods then
+        charmChance = charmChance + charmer:getMod(xi.mod.CHARM_CHANCE)
+    end
+
+    -- apply the dCHR component
+    local dCHR = charmer:getStat(xi.mod.CHR) - target:getStat(xi.mod.CHR)
+    charmChance = charmChance + dCHR
+
+    return utils.clamp(charmChance, 0, 95)
+end
+
+xi.job_utils.beastmaster.attemptCharm = function(charmer, target)
+    if
+        not charmer or        -- Invalid charmer
+        not target or         -- Invalid target
+        not charmer:isPC() or -- Charmer not a player
+        not target:isMob()    -- Target not a mob
+    then
+        return xi.msg.basic.JA_MISS
+    elseif
+        -- Not charmable so apply bind
+        target:getMobMod(xi.mobMod.CHARMABLE) == 0
+    then
+        local resist = applyResistanceAddEffect(charmer, target, xi.element.ICE, 0)
```

Comment:
Styling
```lua
    if
        not charmer or        -- Invalid charmer
        not target or         -- Invalid target
        not charmer:isPC() or -- Charmer not a player
        not target:isMob()    -- Target not a mob
    then
        return xi.msg.basic.JA_MISS

    -- Not charmable so apply bind
    elseif target:getMobMod(xi.mobMod.CHARMABLE) == 0 then
        local resist = applyResistanceAddEffect(charmer, target, xi.element.ICE, 0)
```

Reply:
Thanks for the feedback, I have fixed the PR.